readme.md
## @naturalcycles/scrubber-lib

> Scrub data in JavaScript plain objects by using rules defined in a configuration object

[![npm](https://img.shields.io/npm/v/@naturalcycles/scrubber-lib/latest.svg)](https://www.npmjs.com/package/@naturalcycles/scrubber-lib)
[![Maintainability](https://api.codeclimate.com/v1/badges/e8cd5b1b7cff8e1296fe/maintainability)](https://codeclimate.com/repos/e8cd5b1b7cff8e1296fe/maintainability)
[![Test Coverage](https://api.codeclimate.com/v1/badges/e8cd5b1b7cff8e1296fe/test_coverage)](https://codeclimate.com/repos/e8cd5b1b7cff8e1296fe/test_coverage)
[![code style: prettier](https://img.shields.io/badge/code_style-prettier-ff69b4.svg?style=flat-square)](https://github.com/prettier/prettier)

## How to use

Install it:

```
yarn add -D @naturalcycles/scrubber-lib
```

Define a scrubber configuration object:

```ts
import { ScrubberConfig, Scrubber } from '@naturalcycles/scrubber-lib'

const cfg: ScrubberConfig = {
  fields: {
    name: {
      scrubber: 'staticScrubber',
      params: {
        replacement: 'John Doe',
      },
    },
    password: {
      scrubber: 'undefinedScrubber',
    },
  },
  throwOnError: true, // default: false,
  preserveFalsy: false, // default: true
}
```

Scrub a single object:

```ts
const object = { name: 'Real Name', password: 'secret' }

const scrubber = new Scrubber(cfg)
const scrubbedObject = scrubber.scrub(object)

// scrubbedObject =  name: 'John Doe', password: undefined }
```

Scrub an array of objects:

```ts
const objects = [object1, object2, object3]

const scrubbedObjects = scrubber.scrub(objects)
```

## Public API

```ts
constructor (private cfg: ScrubberConfig, additionalScrubbersImpl?: ScrubbersImpl)
scrub<T> (data: T): T
```

## Features

- **Objects are deep traversed**
- **Immutable** changes (does not mutate the original object)
- TypeScript library, compatible both on browsers and NodeJS
- Fields are scrubbed if object keys match the field names on the configuration file
- Provides a few built-in scrubber functions
- Allows additional scrubber functions
- Validates config object on class initialization to ensure all defined scrubber functions exist
- Supports field names to be comma-separated on configuration file
- Error handling: all errors are logged and allows a `cfg.throwOnError` optional configuration to
  re-throw errors
- Falsy values: allows a `cfg.preserveFalsy` optional configuration to control if falsy values
  should be preserved or passed to scrubber functions. When inspecting scrubbed objects for
  debugging purposes, it might be useful to set it to `true` to identify potential interesting
  fields

## Limitations

- Objects of types `Map`, `Set` and `Buffer` are currently not traversed or modified

## Vocabulary

The `scrubber-lib` supports a `ScrubberConfig` parameter on initialization which is usually defined
by clients on a `scrubber configuration file` (YAML or JSON) with multiple `scrubbing profiles`
(such as anonymization, pseudonymization, etc).

The library applies `scrubber functions` to the given objects. It provides some built-in
`scrubber functions` while also allowing custom `scrubber functions implementations`.

## Possible use cases

Allows, for example, removal of sensitive data for:

- Logs
- Error reporting to third-party services
- Data exports (such as staging or other data exports)
- Anonymizing production users (GDPR "_right to be forgotten_")


CHANGELOG.md
See [Github Releases](https://github.com/NaturalCycles/scrubber-lib/releases)


src/scrubber.test.ts
import { deepFreeze } from '@naturalcycles/test-lib'
import { nanoid } from 'nanoid'
import { Scrubber } from './scrubber'
import { ScrubberConfig, ScrubberFn, ScrubbersImpl } from './scrubber.model'
import { saltedHashEmailScrubber, saltedHashScrubber } from './scrubbers'
import {
  configEmailScrubberMock,
  configInvalidScrubberMock,
  configMultiFieldMock,
  configStaticScrubbersMock,
} from './test/scrubber.mock'

// Convenient method for initializing object and scrubbing
const scrub = <T>(
  data: T,
  cfg: ScrubberConfig = configStaticScrubbersMock(),
  additionalScrubbersImpl?: ScrubbersImpl,
): T => {
  const scrubber = new Scrubber(cfg, additionalScrubbersImpl)
  return scrubber.scrub(data)
}

test('returns a single object when input is a single object', () => {
  const data = { pw: 'secret', name: 'Real Name' }
  const scrubber = new Scrubber(configStaticScrubbersMock())
  const result = scrubber.scrub(data)

  expect(result).toEqual({ pw: 'notsecret', name: 'Jane Doe' })
})

test('returns an array with one object when input is an array with one object', () => {
  const data = { pw: 'secret', name: 'Real Name' }
  const scrubber = new Scrubber(configStaticScrubbersMock())
  const result = scrubber.scrub([data])

  expect(result).toEqual([{ pw: 'notsecret', name: 'Jane Doe' }])
})

test('applies to more than a field', () => {
  const data = [{ pw: 'secret', name: 'Real Name' }]
  deepFreeze(data) // Ensure data doesnt mutate

  const result = scrub(data, configStaticScrubbersMock())
  expect(result).toEqual([{ pw: 'notsecret', name: 'Jane Doe' }])
})

test('applies to nested fields (deep transverse, 2 levels)', () => {
  const data = [{ account: { pw: 'secret', name: 'Real Name' } }]
  deepFreeze(data) // Ensure data doesnt mutate

  const result = scrub(data, configStaticScrubbersMock())
  expect(result).toEqual([{ account: { pw: 'notsecret', name: 'Jane Doe' } }])
})

test('applies to nested fields (deep transverse, 3 levels)', () => {
  const data = [{ object: { account: { pw: 'secret', name: 'Real Name' } } }]
  deepFreeze(data) // Ensure data doesnt mutate

  const result = scrub(data, configStaticScrubbersMock())
  expect(result).toEqual([{ object: { account: { pw: 'notsecret', name: 'Jane Doe' } } }])
})

test('applies to nested arrays', () => {
  const obj1 = { pw: 'shouldChange', safe: 'shouldStay' }
  const obj2 = { name: 'personalInformation', safe2: 'isSafe' }
  const users = [{ users: [obj1, obj2] }]
  deepFreeze(users)

  const result = scrub(users, configStaticScrubbersMock())
  expect(result[0]!['users'][0]).toEqual({ pw: 'notsecret', safe: 'shouldStay' })
  expect(result[0]!['users'][1]).toEqual({ name: 'Jane Doe', safe2: 'isSafe' })
  expect(Array.isArray(result[0]!['users'])).toBeTruthy() // makes sure we don't convert array to objects
})

test('keeps not modified fields', () => {
  const data = [{ safeField: 'keep', email: 'real@email.com' }]
  deepFreeze(data) // Ensure data doesnt mutate

  const result = scrub(data, configEmailScrubberMock())
  expect(result).toEqual([{ safeField: 'keep', email: 'anonymized@email.com' }])
})

test('supports additional scrubbers', () => {
  const mockScrubber: ScrubberFn = () => 'modified'
  const additionalScrubbers: ScrubbersImpl = { aNewScrubber: mockScrubber }

  const cfg: ScrubberConfig = {
    fields: {
      target: {
        scrubber: 'aNewScrubber',
      },
    },
  }

  const data = [{ target: 'original' }]
  deepFreeze(data) // Ensure data doesnt mutate

  const result = scrub(data, cfg, additionalScrubbers)
  expect(result).toEqual([{ target: 'modified' }])
})

test('supports comma-separated fields in field name', () => {
  const data = [{ field1: 'orig1', field2: 'orig2' }]
  deepFreeze(data) // Ensure data doesnt mutate

  const result = scrub(data, configMultiFieldMock())
  expect(result).toEqual([{ field1: 'modified', field2: 'modified' }])
})

test('returns empty array for empty arrays', () => {
  const result = scrub([], configStaticScrubbersMock())
  expect(result).toEqual([])
})

test('fails when scrubber from config is not found (even if not used)', () => {
  expect(() => {
    scrub([], configInvalidScrubberMock())
  }).toThrow()
})

describe('falsy values handling', () => {
  const cfg: ScrubberConfig = {
    fields: {
      target: {
        scrubber: 'undefinedScrubber',
      },
    },
  }

  test.each([-1, true, false, undefined, '', 0, null, NaN])(
    'does not preserve "%s" if configuration is false',
    input => {
      const cfgWithPreserveDisabled: ScrubberConfig = { ...cfg, preserveFalsy: false }

      const scrubber = new Scrubber(cfgWithPreserveDisabled)
      const result = scrubber.scrub({ target: input })

      expect(result.target).toBeUndefined()
    },
  )

  test.each([false, undefined, '', 0, null, NaN])(
    'preserves "%s" if configuration is true',
    input => {
      const cfgWithPreserveEnabled: ScrubberConfig = { ...cfg, preserveFalsy: true }

      const scrubber = new Scrubber(cfgWithPreserveEnabled)
      const result = scrubber.scrub({ target: input })

      expect(result.target).toEqual(input)
    },
  )
})

describe('error handling', () => {
  const faultyScrubber: ScrubberFn = () => {
    throw new Error('ops')
  }

  const cfg: ScrubberConfig = {
    fields: {
      target: {
        scrubber: 'faultyScrubber',
      },
    },
  }

  const object = { target: 'foo' }

  test('logs errors by default', () => {
    jest.spyOn(console, 'log').mockImplementation()
    jest.spyOn(console, 'error').mockImplementation()

    const scrubber = new Scrubber(cfg, { faultyScrubber })
    scrubber.scrub(object)

    expect(console.log).toMatchSnapshot()
    expect(console.error).toMatchSnapshot()
  })

  test('re-throw error if enabled on config', () => {
    const cfgWithErrorsEnabled: ScrubberConfig = { ...cfg, throwOnError: true }

    const scrubber = new Scrubber(cfgWithErrorsEnabled, { faultyScrubber })

    expect(() => {
      scrubber.scrub(object)
    }).toThrow()
  })
})

test('scrubs different types of data', () => {
  const result = scrub([
    {
      null: null,
      undefined, // eslint-disable-line id-blacklist
      array: [1, 2, { pw: 'secret' }],
      function: () => 1,
      symbol: Symbol(42),
      map: new Map([['b', 'c']]),
      set: new Set([1, 2, 3, 4]),
      buffer: Buffer.from('data'),
      date: new Date(0),
    },
  ])

  expect(result).toMatchSnapshot()
})

test('initializationVector is passed as param to scrubbers', () => {
  const mockScrubber = jest.fn(() => 'modified') as any
  const additionalScrubbers: ScrubbersImpl = { aNewScrubber: mockScrubber }

  const cfg: ScrubberConfig = {
    fields: {
      pw: {
        scrubber: 'aNewScrubber',
      },
    },
  }

  const data = [{ pw: 'secret' }]
  deepFreeze(data) // Ensure data doesnt mutate

  const scrubber = new Scrubber(cfg, additionalScrubbers)
  scrubber.scrub(data)

  expect(mockScrubber).toHaveBeenLastCalledWith('secret', {
    initializationVector: expect.any(String),
  })
  scrubber.scrub(data)

  const vector1 = mockScrubber.mock.calls[0][1]
  const vector2 = mockScrubber.mock.calls[1][1]

  expect(vector1).toEqual(vector2)
})

test('Ensure initializationVector is random and affects saltedHashScrubber', () => {
  const mockScrubber = jest.fn(saltedHashScrubber) as any
  const additionalScrubbers: ScrubbersImpl = { aNewScrubber: mockScrubber }

  const cfg: ScrubberConfig = {
    fields: {
      id: {
        scrubber: 'aNewScrubber',
      },
    },
  }

  const toBeSalted = 'id1'
  const data = [{ id: toBeSalted }]
  const scrubber1 = new Scrubber(cfg, additionalScrubbers)
  scrubber1.scrub(data)
  expect(mockScrubber).toHaveBeenCalledWith(toBeSalted, {
    initializationVector: expect.any(String),
  })
  const vector1 = mockScrubber.mock.calls[0][1]
  const result1 = mockScrubber.mock.results[0].value

  const scrubber2 = new Scrubber(cfg, additionalScrubbers) // New vector to be generated by constructor
  scrubber2.scrub(data)
  const vector2 = mockScrubber.mock.calls[1][1]
  const result2 = mockScrubber.mock.results[1].value

  expect(vector1).not.toEqual(vector2)
  expect(result1).not.toEqual(toBeSalted)
  expect(result1).not.toEqual(result2)
})

test('initializationVector passed to scrubber constructor is passed to scrubbers', () => {
  const mockScrubber = jest.fn(saltedHashScrubber) as any
  const additionalScrubbers: ScrubbersImpl = { aNewScrubber: mockScrubber }

  const cfg: ScrubberConfig = {
    fields: {
      id: {
        scrubber: 'aNewScrubber',
      },
    },
  }

  const toBeSalted = 'id1'
  const data = [{ id: toBeSalted }]
  const initializationVector = nanoid()
  const scrubber1 = new Scrubber(cfg, additionalScrubbers, initializationVector)
  scrubber1.scrub(data)
  expect(mockScrubber).toHaveBeenCalledWith(toBeSalted, {
    initializationVector: expect.stringMatching(initializationVector),
  })
  const vector1 = mockScrubber.mock.calls[0][1]
  const result1 = mockScrubber.mock.results[0].value

  const scrubber2 = new Scrubber(cfg, additionalScrubbers, initializationVector)
  scrubber2.scrub(data)
  const vector2 = mockScrubber.mock.calls[1][1]
  const result2 = mockScrubber.mock.results[1].value

  expect(vector1).toEqual(vector2)
  expect(result1).not.toEqual(toBeSalted)
  expect(result1).toEqual(result2)
})

test('supplying an initializationVector in config should take precedence', () => {
  const configVector = '123'
  const mockScrubber = jest.fn(saltedHashScrubber) as any
  const additionalScrubbers: ScrubbersImpl = { aNewScrubber: mockScrubber }

  const cfg: ScrubberConfig = {
    fields: {
      id: {
        scrubber: 'aNewScrubber',
        params: {
          initializationVector: configVector,
        },
      },
    },
  }

  const toBeSalted = 'id1'
  const data = [{ id: toBeSalted }]
  const initializationVector = nanoid()
  const scrubber1 = new Scrubber(cfg, additionalScrubbers, initializationVector)
  const result = scrubber1.scrub(data)

  // Result should stay the same since 123 is used as init vector
  expect(result.pop()).toMatchSnapshot()
})

test('supplying an initializationVector in config of saltedHashEmailScrubber should produce consistent results', () => {
  const configVector = '123'
  const domain = '@example.com.br'
  const mockScrubber = jest.fn(saltedHashEmailScrubber) as any
  const additionalScrubbers: ScrubbersImpl = { aNewScrubber: mockScrubber }

  const cfg: ScrubberConfig = {
    fields: {
      email: {
        scrubber: 'aNewScrubber',
        params: {
          initializationVector: configVector,
          domain,
        },
      },
    },
  }

  const toBeSalted = 'test@email.se'
  const data = [{ email: toBeSalted }]
  const scrubber1 = new Scrubber(cfg, additionalScrubbers)
  const result = scrubber1.scrub(data).pop() as any

  // Result should stay the same since config vector is static
  expect(result).toMatchSnapshot()
  expect(result).not.toEqual(data)
  const emailRes = result['email']
  expect(emailRes).toContain(domain)
  expect(emailRes).not.toContain(toBeSalted)
})

test('example (readme)', () => {
  const cfg: ScrubberConfig = {
    fields: {
      name: {
        scrubber: 'staticScrubber',
        params: {
          replacement: 'John Doe',
        },
      },
      password: {
        scrubber: 'undefinedScrubber',
      },
    },
    throwOnError: true,
  }

  const object = { name: 'Real Name', password: 'secret' }

  const scrubber = new Scrubber(cfg)
  const newObject = scrubber.scrub(object)

  console.log(newObject)
})


src/scrubbers.test.ts
import { nanoid } from 'nanoid'
import {
  bcryptStringScrubber,
  BcryptStringScrubberParams,
  charsFromRightScrubber,
  isoDateStringScrubber,
  preserveOriginalScrubber,
  randomEmailInContentScrubber,
  randomEmailScrubber,
  randomScrubber,
  saltedHashEmailScrubber,
  saltedHashScrubber,
  staticScrubber,
  undefinedScrubber,
  unixTimestampScrubber,
} from './scrubbers'

const bryptStr1 = '$2a$12$HYNzBb8XYOZZeRwZDiVux.orKNqkSVAoXBDc9Gw7nSxr8rcZupbRK'
const bryptStr2 = '$2a$10$HYNzBb8XYOZZeRwZDiVux.orKNqkSVAoXBDc9Gw7nSxr8rcZupbRK'

describe('undefinedScrubber', () => {
  test('replaces any value with undefined', () => {
    expect(undefinedScrubber(true)).toBeUndefined()
    expect(undefinedScrubber(1)).toBeUndefined()
    expect(undefinedScrubber('foo')).toBeUndefined()
  })
})

describe('preserveOriginalScrubber', () => {
  test('preserves the original value', () => {
    expect(preserveOriginalScrubber(true)).toBe(true)
    expect(preserveOriginalScrubber(1)).toBe(1)
    expect(preserveOriginalScrubber('foo')).toBe('foo')
    expect(preserveOriginalScrubber(null)).toBe(null)
  })
})

describe('staticScrubber', () => {
  test.each([
    [undefined, 'replacement'],
    ['', 'replacement'],
  ])('handles undefined values "%s" > "%s"', (input, replacement) => {
    const result = staticScrubber('', { replacement: 'replacement' })
    expect(result).toEqual(replacement)
  })

  test('replaces any string with replacement', () => {
    const o = 'bar'

    expect(staticScrubber('', { replacement: o })).toEqual(o)
    expect(staticScrubber('foo', { replacement: o })).toEqual(o)
  })

  test('replace number', () => {
    const secretNum = 666
    const r = 123

    expect(staticScrubber(secretNum, { replacement: r })).toEqual(r)
  })
})

describe('unixTimestampScrubber', () => {
  test.each([
    [undefined, undefined],
    ['', undefined],
  ])('handles undefined values "%s" > "%s"', (date, expected) => {
    const result = unixTimestampScrubber(date, { excludeDay: true })
    expect(result).toEqual(expected)
  })

  test('scrubs only time (string)', () => {
    // Wednesday, July 3, 2019 9:35:21 AM to
    // Wednesday, July 3, 2019 00:00:00 AM
    const result = unixTimestampScrubber('1562146521', { excludeTime: true })
    expect(result).toEqual(1562112000)
  })

  test('scrubs only time', () => {
    // Wednesday, July 3, 2019 9:35:21 AM to
    // Wednesday, July 3, 2019 00:00:00 AM
    const result = unixTimestampScrubber(1562146521, { excludeTime: true })
    expect(result).toEqual(1562112000)
  })

  test('scrubs only day', () => {
    // Wednesday, July 3, 2019 9:35:21 AM to
    // Wednesday, July 1, 2019 9:35:21 AM
    const result = unixTimestampScrubber(1562146521, { excludeDay: true })
    expect(result).toEqual(1561973721)
  })

  test('scrubs day and month', () => {
    // Wednesday, July 3, 2019 9:35:21 AM to
    // Wednesday, January 1, 2019 9:35:21 AM
    const result = unixTimestampScrubber(1562146521, { excludeDay: true, excludeMonth: true })
    expect(result).toEqual(1546335321)
  })

  test('scrubs only year', () => {
    // Wednesday, July 3, 2019 9:35:21 AM to
    // Wednesday, July 3, 1970 9:35:21 AM
    const result = unixTimestampScrubber(1562146521, { excludeYear: true })
    expect(result).toEqual(15845721)
  })
})

describe('isoDateStringScrubber', () => {
  test.each([
    [undefined, undefined],
    ['', undefined],
  ])('handles undefined values "%s" > "%s"', (date, expected) => {
    const result = isoDateStringScrubber(date, { excludeDay: true })
    expect(result).toEqual(expected)
  })

  test('scrubs only day', () => {
    const result = isoDateStringScrubber('2019-05-12', { excludeDay: true })
    expect(result).toEqual('2019-05-01')
  })

  test('scrubs only month', () => {
    const result = isoDateStringScrubber('2019-05-12', { excludeMonth: true })
    expect(result).toEqual('2019-01-12')
  })

  test('scrubs both day and month', () => {
    const result = isoDateStringScrubber('2019-05-12', { excludeDay: true, excludeMonth: true })
    expect(result).toEqual('2019-01-01')
  })

  test('scrubs only year', () => {
    const result = isoDateStringScrubber('2019-05-12', { excludeYear: true })
    expect(result).toEqual('1970-05-12')
  })

  test('scrubs both day and year', () => {
    const result = isoDateStringScrubber('2019-05-12', { excludeDay: true, excludeYear: true })
    expect(result).toEqual('1970-05-01')
  })
})

describe('charsFromRightScrubber', () => {
  test.each([
    [undefined, undefined],
    ['', undefined],
    ['11225', '112XX'],
    ['ABC DEF', 'ABC DXX'],
    ['ABC', 'AXX'],
    ['AB', 'XX'],
    ['A', 'X'],
  ])('anonymizes zip codes "%s" > "%s"', (zip, expected) => {
    const result = charsFromRightScrubber(zip, { count: 2, replacement: 'X' })
    expect(result).toEqual(expected)
  })

  test('removes 2 chars', () => {
    const result = charsFromRightScrubber('76543', { count: 2, replacement: 'X' })
    expect(result).toEqual('765XX')
  })

  test('does not fail/crash if count > input length', () => {
    const result = charsFromRightScrubber('123', { count: 5, replacement: 'X' })
    expect(result).toEqual('XXX')
  })
})

describe('randomScrubber', () => {
  test('generates with default arguments', () => {
    const result = randomScrubber('secret')
    expect(result).not.toEqual('secret')
  })

  test('accepts length', () => {
    const result = randomScrubber('secret', { length: 5 })
    expect(result).toHaveLength(5)
  })

  test('accepts alphabet and length', () => {
    const result = randomScrubber('secret', { alphabet: 'a', length: 5 })
    expect(result).toEqual('aaaaa')
  })
})

describe('randomEmailScrubber', () => {
  test('generates with default arguments', () => {
    const result = randomEmailScrubber('secret')
    expect(result).not.toEqual('secret')
    expect(result).toContain('@example.com')
  })

  test('accepts alphabet, length and domain', () => {
    const result = randomEmailScrubber('secret', {
      alphabet: 'a',
      length: 5,
      domain: '@customdomain.com',
    })
    expect(result).toEqual('aaaaa@customdomain.com')
  })
})

describe('randomEmailInContentScrubber', () => {
  test('scrub content without email', () => {
    const content = 'I am a string without and email in it @hello!'
    const result = randomEmailInContentScrubber(content)
    expect(result).toEqual(content)
  })

  test('scrub email in URL', () => {
    const email = 'real@gmail.com'
    const prefix = '/api/user/'
    const result = randomEmailInContentScrubber(prefix + email)
    expect(result).not.toContain(email)
    expect(result).toContain(prefix)
  })

  test('scrub complex email in text', () => {
    const email = 'real_email-address.2@gmail2.com.br'
    const suffix = ', not a gmail2.com.br address'
    const text = 'This should be a random email: ' + email + suffix
    const result = randomEmailInContentScrubber(text)

    expect(result).not.toContain(email)
    expect(result).not.toContain('real')
    expect(result).not.toContain('example.com.br') // test multi.dot domains
    expect(result).toContain(suffix)
  })
})

describe('saltedHashScrubber', () => {
  test('generates hash using initializationVector', () => {
    const initializationVector = nanoid()

    const result = saltedHashScrubber('secret', { initializationVector })
    expect(result).not.toEqual('secret')

    const result2 = saltedHashScrubber('secret', { initializationVector })
    expect(result).toEqual(result2)

    const initializationVector2 = nanoid()
    const result3 = saltedHashScrubber('secret', { initializationVector: initializationVector2 })
    expect(result).not.toEqual(result3)
  })
})

describe('saltedHashEmailScrubber', () => {
  test('generates hash using initializationVector and suffixes domain', () => {
    const initializationVector = 'staticvector'

    const result = saltedHashEmailScrubber('secret', {
      initializationVector,
      domain: '@naturalcycles.com',
    })
    console.log(result)
    expect(result).not.toEqual('secret')
    expect(result).toMatchSnapshot()
  })
})

describe('bcryptStringScrubber', () => {
  test('generates valid bcrypt string while maintaining algo and cost factor', () => {
    const result = bcryptStringScrubber(bryptStr1)
    expect(result).not.toEqual(bryptStr1)
    expect(result!.substr(0, 7)).toEqual('$2a$12$')
    expect(result!.length).toEqual(bryptStr1.length)
  })

  test('ensure older cost factor is preserved', () => {
    const result = bcryptStringScrubber(bryptStr2)
    expect(result).not.toEqual(bryptStr2)
    expect(result!.substr(0, 7)).toEqual('$2a$10$')
  })
  test('handling undefined and empty', () => {
    const undefinedResult = bcryptStringScrubber(undefined)
    expect(undefinedResult).toBeUndefined()

    const emptyResult = bcryptStringScrubber('')
    expect(emptyResult).toEqual('')
  })
  test('handling non-valid bcrypt strings, should return valid bcrypt string', () => {
    const result = bcryptStringScrubber('stringWithToFew$')
    expect(result!.substr(0, 7)).toEqual('$2a$12$')
    expect(result!.length).toEqual(bryptStr1.length)
  })
  test('handling replacements map', () => {
    const params: BcryptStringScrubberParams = {
      replacements: '$2a$10$:$2a$10$456,$2a$12$:$2a$12$123',
    }
    const result = bcryptStringScrubber(bryptStr1, params)
    expect(result).toEqual('$2a$12$123')

    const result2 = bcryptStringScrubber(bryptStr2, params)
    expect(result2).toEqual('$2a$10$456')
  })
})


src/scrubbers.ts
import * as crypto from 'crypto'
import { customAlphabet } from 'nanoid'
import { ScrubberFn, ScrubbersImpl } from './scrubber.model'

/*
 Undefined scrubber

 Replace value with `undefined`
 */
export type UndefinedScrubberFn = ScrubberFn<any, undefined>

export const undefinedScrubber: UndefinedScrubberFn = () => undefined

/*
 Preserve original scrubber

 Useful for profiles that inherit from another and want to keep original value
 (eg: "removing" scrubber from parent)
 */
export type PreserveOriginalScrubberFn = ScrubberFn<any, undefined>

export const preserveOriginalScrubber: PreserveOriginalScrubberFn = value => value

/*
 Static scrubber

 Replace value with `params.replacement`
 */
export interface StaticScrubberParams {
  replacement: string | number
}
export type StaticScrubberFn = ScrubberFn<any, StaticScrubberParams>

export const staticScrubber: StaticScrubberFn = (value, params = { replacement: '' }) =>
  params.replacement

/*
 ISO Date string scrubber

 excludeDay: 2019-05-25 -> 2019-05-01
 excludeMonth: 2019-05-25 -> 2019-01-25
 excludeYear: 2019-05-25 -> 1970-05-25
 */
export interface ISODateStringScrubberParams {
  excludeDay?: boolean
  excludeMonth?: boolean
  excludeYear?: boolean
}
export type ISODateStringScrubberFn = ScrubberFn<string | undefined, ISODateStringScrubberParams>

export const isoDateStringScrubber: ISODateStringScrubberFn = (value, params = {}) => {
  if (!value) return

  if (value && params.excludeDay) {
    value = value.substr(0, 8) + '01'
  }

  if (value && params.excludeMonth) {
    value = value.substr(0, 5) + '01' + value.substr(7, 3)
  }

  if (value && params.excludeYear) {
    value = '1970' + value.substr(4, 9)
  }

  return value
}

/*
 Unix timestamp (timestamp in seconds) scrubber
 */
export interface UnixTimestampScrubberParams {
  excludeTime?: boolean
  excludeDay?: boolean
  excludeMonth?: boolean
  excludeYear?: boolean
}
export type UnixTimestampScrubberFn = ScrubberFn<
  number | string | undefined,
  UnixTimestampScrubberParams
>

export const unixTimestampScrubber: UnixTimestampScrubberFn = (value, params = {}) => {
  if (!value) return

  const date = new Date((value as number) * 1000)

  if (value && params.excludeTime) {
    date.setSeconds(0)
    date.setMinutes(0)
    date.setHours(0)
  }

  if (value && params.excludeDay) {
    date.setDate(1)
  }

  if (value && params.excludeMonth) {
    date.setMonth(0)
  }

  if (value && params.excludeYear) {
    date.setFullYear(1970)
  }

  return Math.round(date.getTime() / 1000)
}

/*
  Chars From Right scrubber

  Replace `params.count` characters, from the right to the left, with `params.replacement`
  Useful for anonymizing zip codes
 */
export interface CharsFromRightScrubberParams {
  count: number
  replacement: string
}
export type CharsFromRightScrubberFn = ScrubberFn<string | undefined, CharsFromRightScrubberParams>

export const charsFromRightScrubber: CharsFromRightScrubberFn = (
  value,
  params = { count: 99, replacement: 'X' },
) => {
  if (!value) return

  const { count, replacement } = params

  const lengthToReplace = Math.min(count, value.length)
  return value.substr(0, value.length - count) + replacement.repeat(lengthToReplace)
}

/*
  Random scrubber

  Uses the package nanoid to generate a random string given an alphabet and a length
 */
const ALPHABET_NUMBER = '0123456789'
const ALPHABET_LOWERCASE = 'abcdefghijklmnopqrstuvwxyz'
const ALPHABET_ALPHANUMERIC_LOWERCASE = [ALPHABET_NUMBER, ALPHABET_LOWERCASE].join('')

export interface RandomScrubberParams {
  alphabet?: string
  length?: number
}

export type RandomScrubberFn = ScrubberFn<string, RandomScrubberParams>

export const randomScrubber: RandomScrubberFn = (value, additionalParams) => {
  const params = { alphabet: ALPHABET_ALPHANUMERIC_LOWERCASE, length: 16, ...additionalParams }
  return customAlphabet(params.alphabet, params['length'])()
}

/*
  Random email scrubber

  Uses the package nanoid to generate a random string given an alphabet and a length
  and appends a given domain (should include '@') at the end of it
 */
export interface RandomEmailScrubberParams {
  alphabet?: string
  length?: number
  domain?: string
}

export type RandomEmailScrubberFn = ScrubberFn<string, RandomEmailScrubberParams>

export const randomEmailScrubber: RandomEmailScrubberFn = (value, additionalParams) => {
  const params = {
    alphabet: ALPHABET_ALPHANUMERIC_LOWERCASE,
    length: 16,
    domain: '@example.com',
    ...additionalParams,
  }
  return customAlphabet(params.alphabet, params['length'])() + params.domain
}

/*
  Random email in content scrubber

  Extends the random email scrubber and allows scrubbing emails within strings while maintaining the rest of the string
 */
export interface RandomEmailInContentScrubberParams {
  alphabet?: string
  length?: number
  domain?: string
}

export type RandomEmailInContentScrubberFn = ScrubberFn<string, RandomEmailInContentScrubberParams>

export const randomEmailInContentScrubber: RandomEmailInContentScrubberFn = (
  value,
  additionalParams,
) => {
  // Email regex, allows letters
  // eslint-disable-next-line no-useless-escape
  const emailRegex = /([a-zA-Z1-9\._-]*@[a-zA-Z1-9\._-]*\.[a-zA-Z_-]{2,3})/
  const matches = emailRegex.exec(value)
  if (!matches) {
    // No email found, return as is
    return value
  } else {
    // Replace all matches with random email
    const match = matches.pop() as string
    value = value.replace(match, randomEmailScrubber(value, additionalParams))

    return value
  }
}

/*
  Salted hash scrubber.

  Takes an initializationVector param and uses it to salt the value before hashing it.
 */
export interface SaltedHashScrubberParams {
  initializationVector: string
}

export type SaltedHashScrubberFn = ScrubberFn<string, SaltedHashScrubberParams>

export const saltedHashScrubber: SaltedHashScrubberFn = (value, params) => {
  if (!params || !params.initializationVector) {
    throw new Error('Initialization vector is missing')
  }

  return crypto.createHash('sha256').update(value).update(params.initializationVector).digest('hex')
}

/*
  Salted hash email scrubber.

  Takes an initializationVector param and uses it to salt the value before hashing it and suffixing email domain
 */
export interface SaltedHashEmailScrubberParams {
  initializationVector: string
  domain?: string
}

export type SaltedHashEmailScrubberFn = ScrubberFn<string, SaltedHashEmailScrubberParams>

export const saltedHashEmailScrubber: SaltedHashEmailScrubberFn = (value, additionalParams) => {
  const params = {
    domain: '@example.com',
    ...additionalParams,
  } as SaltedHashEmailScrubberParams

  if (!params || !params.initializationVector) {
    throw new Error('Initialization vector is missing')
  }

  return saltedHashScrubber(value, params) + params.domain
}

/*
 Bcrypt string scrubber. Scrubs both salt and hash while maintaining algo and cost factor, thus resulting in a valid,
 but nonsense bcrypt string

 */
export type BcryptStringScrubberFn = ScrubberFn<string | undefined, BcryptStringScrubberParams>

/*
 replacements string is a comma seperated list of key-value pairs (seperated by :) that maps bcrypt string prefix
 (algo + cost factor) to a resulting string replacement.

 e.g. replacements: '$2a$10$:$2a$10$456,$2a$12$:$2a$12$123'
 */
export interface BcryptStringScrubberParams {
  replacements: string
}

export const bcryptStringScrubber: BcryptStringScrubberFn = (value, params) => {
  if (!value) return value

  // Keep value until 3rd $
  const cutoff = nthChar(value, '$', 3)
  if (!cutoff) return `$2a$12$${customAlphabet(ALPHABET_ALPHANUMERIC_LOWERCASE, 53)()}`

  const prefix = value.substring(0, cutoff)

  if (params?.replacements) {
    for (const kvPair of params.replacements.split(',')) {
      const [k, v] = kvPair.split(':')
      if (prefix === k) return v
    }
  }

  return `${prefix}${customAlphabet(ALPHABET_ALPHANUMERIC_LOWERCASE, 53)()}`
}

function nthChar(str: string, character: string, n: number): number | undefined {
  let count = 0
  let i = 0
  while (count < n) {
    i = str.indexOf(character, i) + 1
    if (i < 1) {
      return
    }
    count++

    if (count === n) return i
  }
}

export const defaultScrubbers: ScrubbersImpl = {
  staticScrubber,
  preserveOriginalScrubber,
  isoDateStringScrubber,
  unixTimestampScrubber,
  undefinedScrubber,
  charsFromRightScrubber,
  randomScrubber,
  randomEmailScrubber,
  randomEmailInContentScrubber,
  saltedHashScrubber,
  saltedHashEmailScrubber,
  bcryptStringScrubber,
}
